packages/unlock-js/src/PublicLock/v10/purchaseKeys.js

```diff
@@ -163,14 +101,41 @@ export default async function (
     }
   }
 
-  const transactionRequest = await lockContract.populateTransaction.purchase(
-    keyPrices,
-    owners,
-    referrers,
-    keyManagers,
-    data,
-    transactionOptions
-  )
+  // If the lock is priced in ERC20, we need to approve the transfer
+  const approvalOptions = swap
+    ? {
+        erc20Address: swap.srcTokenAddress,
+        address: unlockSwapPurchaserContract?.address,
+        totalAmountToApprove,
+      }
+    : {
+        erc20Address,
+        totalAmountToApprove,
+        address: lockAddress,
+      }
+
+  await approveAllowance.bind(this)(approvalOptions)
+
+  const transactionRequestPromise = swap
+    ? unlockSwapPurchaserContract?.populateTransaction?.swapAndCall(
+        lockAddress,
+        swap.srcTokenAddress,
+        swap.amountInMax,
+        swap.uniswapRouter,
+        swap.swapCallData,
+        callData,
+        transactionOptions
+      )
+    : lockContract.populateTransaction.purchase(
```

Comment:
I wonder if we could just pass `callData` here! 
```suggestion
    : lockContract.populateTransaction.purchase( // TODO consider passing `callData` directly!
```